File: src/main/java/com/danyazero/JuneVisitor.java
package com.danyazero;

import com.danyazero.model.Operation;
import com.danyazero.model.ast.*;
import com.danyazero.utils.*;
import june.GoParser;
import june.GoParserBaseVisitor;
import org.antlr.v4.runtime.tree.ParseTree;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class JuneVisitor extends GoParserBaseVisitor<Node> {
    @Override
    public Node visitSourceFile(GoParser.SourceFileContext ctx) {
        return visitObjectDeclaration(ctx.objectDeclaration());
    }

    @Override
    public NodeList visitBlock(GoParser.BlockContext ctx) {
        var statementList = new ArrayList<Node>();
        for (var s : ctx.statementList().statement()) {
            statementList.add(visit(s));
        }

        return new NodeList(statementList);
    }

    @Override
    public Node visitClassBodyDeclaration(GoParser.ClassBodyDeclarationContext ctx) {
        if (ctx.functionDecl() != null) {
            return visitFunctionDecl(ctx.functionDecl());
        } else if (ctx.methodDecl() != null) {
            return visit(ctx.methodDecl());
        } else if (ctx.classFields() != null) {
            return visit(ctx.classFields());
        }

        throw new RuntimeException("Illegal ClassBodyDeclaration");
    }

    @Override
    public JuneClass visitObjectDeclaration(GoParser.ObjectDeclarationContext ctx) {
        String name = ctx.classDeclaration().identifier().getText();


        var members = new ArrayList<Node>();
        for (var el : ctx.classDeclaration().classBody().classBodyDeclaration()) {
            members.add(visitClassBodyDeclaration(el));
        }

//        var implementsTypes = new ArrayList<Node>();
        if (ctx.classDeclaration().IMPLEMENTS() != null) {
//            for (var typeCtx : ctx.classDeclaration().typeList().type_()) {
//                implementsTypes.add(visitType_(typeCtx));
//            }

            return new JuneClass(name, members);

//            visitAnnotationList(ctx.annotationList()), // Way to get annotation list
//            Arrays.toString(implementsTypes.toArray()), // Way to get types which this class implements
        } else if (ctx.classDeclaration().EXTENDS() != null) {
            return new JuneClass(name, members);
//            ctx.classDeclaration().type_().typeName().getText(), //Way to get type which this class extends
        }


        return new JuneClass(name, members);
    }

    @Override
    public Node visitModifier(GoParser.ModifierContext ctx) {
        if (ctx.PUBLIC() != null) {
            return null; //TODO Provide modifier support
        }

        throw new RuntimeException("Illegal Modifier");
    }

    @Override
    public Node visitAnnotation(GoParser.AnnotationContext ctx) {

        var params = new ArrayList<String>();
        if (ctx.normalAnnotation().elementValuePairList() != null) {
            for (var param : ctx.normalAnnotation().elementValuePairList().elementValuePair()) {
                params.add(String.format("Pair { name=%s, value=%s }", param.identifier(), visit(param.elementValue().basicLit())));
            }
        }

//        return String.format("Annotation { name=%s, values=%s }", ctx.normalAnnotation().typeName().IDENTIFIER(), Arrays.toString(params.toArray()));
        throw new RuntimeException("Not implemented yet."); //TODO Annotation visitor
    }

    @Override
    public Node visitAnnotationList(GoParser.AnnotationListContext ctx) {
        throw new RuntimeException("Not implemented yet."); //TODO Annotation list

//        var annotations = new ArrayList<Node>();
//        for (var annotation : ctx.annotation()) {
//            annotations.add(visitAnnotation(annotation));
//        }

//        return Arrays.toString(annotations.toArray());
    }

    @Override
    public TypeNode visitType_(GoParser.Type_Context ctx) {
        if (ctx.typeName() != null) {
            var typeObj = getType(ctx.typeName().getText());

            return new TypeNode(typeObj);
        } else if (ctx.typeLit() != null) {
            return visitTypeLit(ctx.typeLit());
        }

        throw new RuntimeException("Illegal Type");
    }

    @Override
    public TypeNode visitTypeLit(GoParser.TypeLitContext ctx) {
        if (ctx.arrayType() != null) {
            var type = visitType_(ctx.arrayType().elementType().type_());
            if (type instanceof TypeNode typeNode) {
                return new TypeNode(new ArrayType(typeNode.getType()));
            }

            throw new RuntimeException("Incorrect type format");
        }

        throw new RuntimeException("Illegal Lit Type");
    }

    @Override
    public ReturnStatement visitReturnStmt(GoParser.ReturnStmtContext ctx) {
        return new ReturnStatement(visit(ctx.expressionList()));
    }

    @Override
    public TypeList visitSignature(GoParser.SignatureContext ctx) {
        var resultParameters = new ArrayList<TypeNode>();
        if (ctx.result() != null) {
            if (ctx.result().type_() != null) {
                var type = visitType_(ctx.result().type_());
                resultParameters.add(type);
            } else if (ctx.result().parameters() != null) {
                for (var param : ctx.result().parameters().parameterDecl()) {
                    var type = visitTypeLit(param.type_().typeLit());
                    resultParameters.add(type);
                }
            }
        }

        return new TypeList(resultParameters);
    }

    @Override
    public Method visitFunctionDecl(GoParser.FunctionDeclContext ctx) {
        List<MethodParameter> params = new ArrayList<>();
        if (ctx.signature() != null) {
            var paramCtx = ctx.signature().parameters();
            if (paramCtx != null) {
                for (var p : paramCtx.parameterDecl()) {
                    String paramName = p.identifierList().getText();
//                    String type = p.type_() != null ?  : ;
                    var parameter = new MethodParameter(paramName, visitType_(p.type_()).getType());
                    params.add(parameter);
                }
            }
        }

        return new Method(
                ctx.IDENTIFIER().getText(),
                params,
                visitBlock(ctx.block()).getStatements(),
                visitSignature(ctx.signature()).getStatements()
        );
    }

    @Override
    public Node visitStatementList(GoParser.StatementListContext ctx) {
        var blocks = new ArrayList<Node>();
        for (var el : ctx.statement()) {
            blocks.add(visit(el));
        }

//        return Arrays.toString(blocks.toArray());
        return null;
    }

    //TODO Fix visitIfStatement method
//    @Override
//    public Node visitIfStmt(GoParser.IfStmtContext ctx) {
//        var statementExpression = visitExpression(ctx.expression());
//
//        var statementBlock = new ArrayList<String>();
//        for (var block : ctx.block()) {
//            statementBlock.add(visit(block.statementList()));
//        }
//
//
//        return String.format("IfStmt { expression=%s , block=%s }", statementExpression, Arrays.toString(statementBlock.toArray()));
//    }


    @Override
    public Node visitShortVarDecl(GoParser.ShortVarDeclContext ctx) {
        var expression = visitExpression(ctx.expressionList().expression(0));

        return new Variable(
                ctx.identifierList().getText(),
                expression.getType(),
                expression
        );
    }

    @Override
    public Node visitExpressionStmt(GoParser.ExpressionStmtContext ctx) {
        return visitChildren(ctx);
    }

    @Override
    public Node visitConstSpec(GoParser.ConstSpecContext ctx) {
        throw new RuntimeException("Not implemented yet."); //TODO Implement constant variable
//        return String.format(
//                "Const { name=%s, type=%s, value=%s }",
//                ctx.identifierList().getText(),
//                visitType_(ctx.type_()),
//                visitExpression(ctx.expressionList().expression(0))
//        );
    }

    @Override
    public Node visitVarSpec(GoParser.VarSpecContext ctx) {

        return new Variable(
                ctx.identifierList().getText(),
                visitType_(ctx.type_()).getType(),
                visitExpression(ctx.expressionList().expression(0))
        );
    }

    @Override
    public Expression visitOperand(GoParser.OperandContext ctx) {
        if (ctx.literal() != null) {
            return visitLiteral(ctx.literal());
        } else if (ctx.operandName() != null) {
            if (ctx.operandName().getText().equals("true") || ctx.operandName().getText().equals("false")) {
                return new Value<>(Boolean.parseBoolean(ctx.operandName().getText()), new BooleanType());
            }

            return new Operand(ctx.operandName().getText());
        } else if (ctx.expression() != null) {
            return visitExpression(ctx.expression());
        }

        throw new RuntimeException("Provided wrong expression or operand");
    }

    @Override
    public Expression visitExpression(GoParser.ExpressionContext ctx) {
        if (ctx.expression().size() == 2) {
            var left = visitExpression(ctx.expression(0));
            var right = visitExpression(ctx.expression(1));
            Operation op = null;

            for (ParseTree child : ctx.children) {
                String text = child.getText();
                if (!text.isBlank() && !text.equals("(") && !text.equals(")")) {
                    if (!(child instanceof GoParser.ExpressionContext)) {
                        op = Operation.fromString(text);
                        break;
                    }
                }
            }

            return new ExpressionNode(op, left, right);
        }

        if (ctx.unary_op != null) {
            String op = ctx.unary_op.getText();
            var inner = visitExpression(ctx.expression(0));

            throw new RuntimeException("Not implemented yet."); //TODO Implement Unary operation processing
//            return String.format("UnaryOperation { operation=%s, inner=%s }", op, inner);
        } else if (ctx.primaryExpr() != null) {
//            throw new RuntimeException("Not implemented yet."); //TODO Implement primaryExpression
            return visitPrimaryExpr(ctx.primaryExpr());
        } else if (ctx.shortSliceDecl() != null) {
            return visitShortSliceDecl(ctx.shortSliceDecl());
        }

        throw new RuntimeException("Unhandled expression: " + ctx.getText());
    }

    @Override
    public Node visitAssignment(GoParser.AssignmentContext ctx) {
        var operation = ctx.assign_op();
        throw new RuntimeException("Not implemented yet."); //TODO Implement assignment

//        if (operation.ASSIGN() != null) {
//            var left = visit(ctx.expressionList(0).expression(0));
//            var right = visit(ctx.expressionList(1).expression(0));
//
//            if (operation.PLUS() != null) {
//                var increment = String.format("Increment { operand=%s, value=%s }", left, right);
//
//                return String.format("Assignment { operand=%s, expr=%s }", left, increment);
//            } else if (operation.MINUS() != null) {
//                var increment = String.format("Increment { operand=%s, value=Negate { value=%s } }", left, right);
//
//                return String.format("Assignment { operand=%s, expr=%s }", left, increment);
//            } else if (operation.STAR() != null) {
//                var multiplication = String.format("Expression { left=%s, op=*, right=%s}", left, right);
//
//                return String.format("Assignment { operand=%s, expr=%s }", left, multiplication);
//            } else if (operation.DIV() != null) {
//                var division = String.format("Expression { left=%s, op=/, right=%s}", left, right);
//
//                return String.format("Assignment { operand=%s, expr=%s }", left, division);
//            } else if (operation.MOD() != null) {
//                var mod = String.format("Expression { left=%s, op=%%, right=%s}", left, right);
//
//                return String.format("Assignment { operand=%s, expr=%s }", left, mod);
//            } else if (operation.RSHIFT() != null) {
//                var rightShift = String.format("Expression { left=%s, op=>>, right=%s}", left, right);
//
//                return String.format("Assignment { operand=%s, expr=%s }", left, rightShift);
//            } else if (operation.LSHIFT() != null) {
//                var leftShift = String.format("Expression { left=%s, op=<<, right=%s}", left, right);
//
//                return String.format("LShiftAssignment { operand=%s, expr=%s }", left, leftShift);
//            } else {
//                return String.format("Assignment { operand=%s, expr=%s }", left, right);
//            }
//        }
//
//        throw new RuntimeException("Unhandled assignment: " + ctx.getText());
    }

    @Override
    public ExpressionList visitArguments(GoParser.ArgumentsContext ctx) {
        var arguments = new ArrayList<Expression>();
        for (var el : ctx.expressionList().expression()) {
            arguments.add(visitExpression(el));
        }
        return new ExpressionList(arguments);
    }

    @Override
    public Expression visitPrimaryExpr(GoParser.PrimaryExprContext ctx) {
        if (ctx.methodExpr() != null) {
            var object = ctx.methodExpr().type_().typeName().getText();
            var method = ctx.methodExpr().IDENTIFIER().getText();

            var arguments = new ArrayList<ExpressionList>();
            for (var argument : ctx.arguments()) {
                arguments.add(visitArguments(argument));
            }
            System.out.println("Method arguments: " + arguments);

//            throw new RuntimeException("Method expression not implemented yet");

            return new MethodInvoke(
                    object + "." + method,
                    arguments.getFirst().getExpressions()
            );
//            return String.format(
//                    "MethodInvoke { method=%s, params=%s }",
//                    object + "." + method,
//                    Arrays.toString(arguments.toArray())
//            );
        } else if (ctx.operand() != null && ctx.index() != null && !ctx.index().isEmpty()) {
            return new SliceElement(
                    visitOperand(ctx.operand()),
                    visitOperand(ctx.index(0).expression().primaryExpr().operand())
            );
        }

        return (Expression) visitChildren(ctx);
//        throw new RuntimeException("This kind of primary expression is not supported");
    }

    @Override
    public Value<?> visitLiteral(GoParser.LiteralContext ctx) {
        if (ctx.basicLit() != null) {
            if (ctx.basicLit().string_() != null) {
                String raw = ctx.basicLit().string_().getText();
                String value = raw.substring(1, raw.length() - 1);

                return new Value<>(value, new StringType());
            } else if (ctx.basicLit().integer() != null) {
                return new Value<>(Integer.parseInt(ctx.basicLit().integer().getText()), new IntegerType());
            } else if (ctx.basicLit().getText().contains(".")) {
                return new Value<>(Double.parseDouble(ctx.basicLit().getText()), new DoubleType());
            }
        }

        throw new RuntimeException("Unhandled literal: " + ctx.getText());
    }

    @Override
    public Expression visitShortSliceDecl(GoParser.ShortSliceDeclContext ctx) {
        var items = new ArrayList<Expression>();
        for (var item : ctx.literalList().expression()) {
            items.add(visitExpression(item));
        }

        return new Slice(items);
    }

    private static com.danyazero.model.Type<?> getType(String type) {
        System.out.println("getType(" + type + ");");
        var res = switch (type) {
            case "int" -> new IntegerType();
            case "bool" -> new BooleanType();
            case "string" -> new StringType();
            default -> new ObjectType(null); //TODO Provide correct object name
        };
        System.out.println("return " + res.getClass());

        return res;
    }
}

File: src/main/kotlin/JuneVisitor.kt
package com.danyazero

import com.danyazero.expression.*
import com.danyazero.model.Expression
import com.danyazero.model.Node
import com.danyazero.model.ToRange
import com.danyazero.model.Type
import com.danyazero.model.UntilRange
import com.danyazero.node.*
import com.danyazero.node.Array
import com.danyazero.type.*
import june.JuneParser
import june.JuneParserBaseVisitor
import org.objectweb.asm.Opcodes

class JuneVisitor : JuneParserBaseVisitor<Node>() {
    override fun visitSourceFile(ctx: JuneParser.SourceFileContext): Node {
        return visitClassDeclaration(ctx.classDeclaration())
    }

    override fun visitClassDeclaration(ctx: JuneParser.ClassDeclarationContext): Node {
        return Class(
            name = ctx.identifier().text,
            members = visitClassBody(ctx.classBody()).nodes
        )
    }

    override fun visitClassBody(ctx: JuneParser.ClassBodyContext): NodeList {
        return NodeList(ctx.classBodyDeclaration().map { visitClassBodyDeclaration(it) })
    }

    override fun visitClassBodyDeclaration(ctx: JuneParser.ClassBodyDeclarationContext): Node {
        if (ctx.functionDecl() != null) {
            return visitFunctionDecl(ctx.functionDecl())
        }

        throw RuntimeException("Unsupported class member")
    }

    override fun visitFunctionDecl(ctx: JuneParser.FunctionDeclContext): Node {
        val resultType = if (ctx.signature().result() != null) {
            visitResult(ctx.signature().result()).types
        } else {
            listOf()
        }
        var modifiers = ctx.modifier().map {
            when  {
                it.PUBLIC() != null -> Opcodes.ACC_PUBLIC
                else -> Opcodes.ACC_PRIVATE
            }
        }.sum()

        if (ctx.STAR() != null) {
            modifiers += Opcodes.ACC_STATIC
        }

        return Method(
            name = ctx.IDENTIFIER().text,
            modifiers = modifiers,
            returnTypes = resultType,
            parameters = visitParameters(ctx.signature().parameters()).parameters,
            statementList = visitBlock(ctx.block()).nodes
        )
    }

    override fun visitBlock(ctx: JuneParser.BlockContext): NodeList {
        return NodeList(ctx.statementList().statement().map { visit(it) })
    }

    override fun visitResult(ctx: JuneParser.ResultContext): TypeList {
        return if (ctx.type_() != null) {
            TypeList(listOf(visitType_(ctx.type_())))
        } else if (ctx.results() != null) {
            TypeList(ctx.results().type_().map { visitType_(it) })
        } else throw RuntimeException("Unexpected signature")
    }

    override fun visitParameters(ctx: JuneParser.ParametersContext): ParameterList {
        return ParameterList(ctx.parameterDecl().map { visitParameterDecl(it) })
    }

    override fun visitParameterDecl(ctx: JuneParser.ParameterDeclContext): Parameter {
        return Parameter(name = ctx.IDENTIFIER().text, type = visitType_(ctx.type_()).type)
    }

    override fun visitVarSpec(ctx: JuneParser.VarSpecContext): Node {
        return Variable(
            name = ctx.identifierList().getText(),
            type = if (ctx.type_() != null) visitType_(ctx.type_()).type else null,
            value = visitExpression(ctx.expressionList().expression(0)),
        )
    }

    override fun visitConstSpec(ctx: JuneParser.ConstSpecContext): Node {
        return Variable(
            name = ctx.identifierList().getText(),
            type = if (ctx.type_() != null) visitType_(ctx.type_()).type else null,
            value = visitExpression(ctx.expressionList().expression(0)),
            isConstant = true
        )
    }

    override fun visitShortVarDecl(ctx: JuneParser.ShortVarDeclContext): Node {
        return Variable(
            name = ctx.identifierList().getText(),
            value = visitExpression(ctx.expressionList().expression(0))
        )
    }

    override fun visitType_(ctx: JuneParser.Type_Context): TypeNode {
        if (ctx.IDENTIFIER() != null) {
            return TypeNode(getType(ctx.IDENTIFIER().text))
        } else if (ctx.arrayType() != null) {
            return TypeNode(ArrayType(visitType_(ctx.arrayType().type_()).type))
        }

        throw RuntimeException("Illegal Type")
    }

    override fun visitMacroStmt(ctx: JuneParser.MacroStmtContext): Node {
        return if (ctx.IDENTIFIER().text == "print") {
            PrintMacros(visitExpression(ctx.arguments().expressionList().expression(0)))
        } else throw RuntimeException("Unsupported macros")
    }

    override fun visitExpressionList(ctx: JuneParser.ExpressionListContext): ExpressionList {
        return ExpressionList(ctx.expression().map { visitExpression(it) })
    }

    override fun visitExpression(ctx: JuneParser.ExpressionContext): Expression {
        if (ctx.expression().size == 2) {
            val left = visitExpression(ctx.expression(0))
            val right = visitExpression(ctx.expression(1))

            val expression = when {
                ctx.PLUS() != null -> ::AdditionExpression
                ctx.STAR() != null -> ::MultiplicationExpression
                ctx.DIV() != null -> ::DivisionExpression
                ctx.MINUS() != null -> ::SubstructionExpression
                ctx.LSHIFT() != null -> ::ShiftLeftExpression
                ctx.RSHIFT() != null -> ::ShiftRightExpression
                ctx.LRSHIFT() != null -> ::LogicalShiftRightExpression
                ctx.MOD() != null -> ::RemainderExpression
                ctx.EQUALS() != null -> ::EqualExpression
                ctx.LESS() != null -> ::LessExpression
                ctx.GREATER() != null -> ::GreaterExpression
                ctx.LESS_OR_EQUALS() != null -> ::LessOrEqualExpression
                ctx.GREATER_OR_EQUALS() != null -> ::GreaterOrEqualExpression
                else -> throw RuntimeException("Unsupported expression operand")
            }

            return expression(left, right)
        } else if (ctx.arrayDecl() != null) {
            return visitArrayDecl(ctx.arrayDecl())
        } else if (ctx.primaryExpr() != null) {
            return visitPrimaryExpr(ctx.primaryExpr())
        }

        throw RuntimeException("Unsupported expression")
    }

    override fun visitAssignment(ctx: JuneParser.AssignmentContext): AssignExpression {
        val operation = ctx.assign_op()

        if (operation.ASSIGN() != null) {
            val operand = visitExpression(ctx.expression(0))
            if (operand is Operand) {
                val right = visitExpression(ctx.expression(1))


                val expression = when {
                    operation.PLUS() != null -> ::AdditionExpression
                    operation.MINUS() != null -> ::SubstructionExpression
                    operation.STAR() != null -> ::MultiplicationExpression
                    operation.DIV() != null -> ::DivisionExpression
                    operation.MOD() != null -> ::RemainderExpression
                    else -> null
                }

                if (expression != null) {
                    return AssignExpression(operand = operand, expression = expression(operand, right))
                }

                return AssignExpression(operand = operand, expression = right)
            }

            throw RuntimeException("Left expression should be Operand { }: " + ctx.getText())
        }


        throw RuntimeException("Unhandled assignment: " + ctx.getText())
    }

    override fun visitLoopStmt(ctx: JuneParser.LoopStmtContext): Node {
        if (ctx.expression() != null && ctx.expression().rel_op != null) {
            val conditionExpression = visitExpression(ctx.expression())

            if (conditionExpression is ConditionExpression) {
                return Loop(condition = conditionExpression, statements = visitBlock(ctx.block()).nodes)
            } else {
                throw RuntimeException("Unhandled loop condition expression")
            }
        }

        return Loop(statements = visitBlock(ctx.block()).nodes)
    }

    override fun visitRangeLoopStmt(ctx: JuneParser.RangeLoopStmtContext): Node {


        val left = visitExpression(ctx.range().expression(0))
        val right = visitExpression(ctx.range().expression(1))

        val range = when {
            ctx.range().UNTIL() != null -> ::UntilRange
            ctx.range().TO() != null -> ::ToRange
            else -> throw RuntimeException("Unhandled range expression")
        }

        return RangeLoop(
            index = ctx.IDENTIFIER().text,
            range = range(left, right),
            statements = visitBlock(ctx.block()).nodes
        )
    }

    override fun visitOperandLoopStmt(ctx: JuneParser.OperandLoopStmtContext): Node {
        if (ctx.expression() != null) {
            val operand = visitExpression(ctx.expression())

            return if (ctx.loopParameters() == null) {
                OperandLoop(operand = operand, statements = visitBlock(ctx.block()).nodes)
            } else if (ctx.loopParameters().IDENTIFIER().size == 2) {
                OperandLoop(
                    operand = operand,
                    index = ctx.loopParameters().IDENTIFIER(0).text,
                    item = ctx.loopParameters().IDENTIFIER(1).text,
                    statements = visitBlock(ctx.block()).nodes
                )
            } else {
                OperandLoop(operand = operand, item = ctx.loopParameters().IDENTIFIER(0).text, statements = visitBlock(ctx.block()).nodes)
            }
        }

        throw RuntimeException("Unsupported loop")
    }

    override fun visitIncDecStmt(ctx: JuneParser.IncDecStmtContext): Node {
        val operand = visitExpression(ctx.expression())
        if (operand !is Operand) throw RuntimeException("Unhandled operand type.")

        return when {
            ctx.PLUS_PLUS() != null -> AssignExpression(
                operand = operand,
                expression = AdditionExpression(left = operand, right = ValueExpression.of(1))
            )

            ctx.MINUS_MINUS() != null -> AssignExpression(
                operand = operand,
                expression = SubstructionExpression(left = operand, right = ValueExpression.of(1))
            )

            else -> throw RuntimeException("Unsupported expression operand")
        }
    }

    override fun visitArrayDecl(ctx: JuneParser.ArrayDeclContext): Array {
        val expressionList = visitExpressionList(ctx.expressionList())

        return Array(
            items = expressionList.items
        )
    }

    override fun visitPrimaryExpr(ctx: JuneParser.PrimaryExprContext): Expression {
        if (ctx.operand() != null && ctx.index() != null && !ctx.index().isEmpty()) {
            return ArrayValue(
                visitOperand(ctx.operand()),
                visitExpression(ctx.index(0).expression())
            )
        } else if (ctx.operand() != null) {
            return visitOperand(ctx.operand())
        }

        throw RuntimeException("Unsupported primary expression")
    }

    override fun visitOperand(ctx: JuneParser.OperandContext): Expression {
        if (ctx.literal() != null) {
            return visitLiteral(ctx.literal())
        } else if (ctx.operandName() != null) {
            if (ctx.operandName().text == "true" || ctx.operandName().text == "false") {
                return ValueExpression(
                    ctx.operandName()?.text.toBoolean(),
                    BooleanType()
                )
            }

            return Operand(ctx.operandName().getText())
        } else if (ctx.expression() != null) {
            return visitExpression(ctx.expression())
        }

        throw RuntimeException("Provided wrong expression or operand");
    }

    override fun visitLiteral(ctx: JuneParser.LiteralContext): Expression {
        if (ctx.basicLit() != null) {
            if (ctx.basicLit().string_() != null) {
                val raw = ctx.basicLit().string_().getText()
                val value = raw.substring(1, raw.length - 1)

                return ValueExpression(value, StringType())
            } else if (ctx.basicLit().integer() != null) {
                return ValueExpression(Integer.parseInt(ctx.basicLit().integer().text), IntegerType())
            } else if (ctx.basicLit().getText().contains(".")) {
                return ValueExpression(ctx.basicLit().text.toDouble(), DoubleType())
            }
        }

        throw RuntimeException("Unhandled literal: " + ctx.getText())
    }

    private fun getType(type: String): Type<*> {
        return when (type) {
            "int" -> IntegerType()
            "bool" -> BooleanType()
            "string" -> StringType()
            else -> ObjectType(null)
        }
    }

}